core/contracts/types/entitlement_types.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func (o *OrOperation) GetRightOperation() Operation {
+	return o.RightOperation
+}
+
+func (o *OrOperation) SetLeftOperation(left Operation) {
+	o.LeftOperation = left
+}
+
+func (o *OrOperation) SetRightOperation(right Operation) {
+	o.RightOperation = right
+}
+
+type AndOperation struct {
+	LogicalOperation // Embedding LogicalOperation interface
+	OpType           OperationType
+	LogicalType      LogicalOperationType
+	LeftOperation    Operation
+	RightOperation   Operation
+}
+
+func (a *AndOperation) GetOpType() OperationType {
+	return a.OpType
+}
+
+func (a *AndOperation) GetLogicalType() LogicalOperationType {
+	return a.LogicalType
+}
+
+func (a *AndOperation) GetLeftOperation() Operation {
+	return a.LeftOperation
+}
+
+func (a *AndOperation) GetRightOperation() Operation {
+	return a.RightOperation
+}
+
+func (a *AndOperation) SetLeftOperation(left Operation) {
+	a.LeftOperation = left
+}
+
+func (a *AndOperation) SetRightOperation(right Operation) {
+	a.RightOperation = right
+}
+
+func GetOperationTree(
+	ctx context.Context,
+	ruleData *base.IRuleEntitlementBaseRuleDataV2,
+) (Operation, error) {
+	log := dlog.FromCtx(ctx)
+	decodedOperations := []Operation{}
+	log.Debug("Decoding operations", "ruleData", ruleData)
+	for _, operation := range ruleData.Operations {
+		if OperationType(operation.OpType) == CHECK {
+			checkOperation := ruleData.CheckOperations[operation.Index]
+			decodedOperations = append(decodedOperations, &CheckOperation{
+				OpType:          CHECK,
+				CheckType:       CheckOperationType(checkOperation.OpType),
+				ChainID:         checkOperation.ChainId,
+				ContractAddress: checkOperation.ContractAddress,
+				Params:          checkOperation.Params,
+			})
+		} else if OperationType(operation.OpType) == LOGICAL {
+			logicalOperation := ruleData.LogicalOperations[operation.Index]
+			if LogicalOperationType(logicalOperation.LogOpType) == AND {
+				decodedOperations = append(decodedOperations, &AndOperation{
+					OpType:         LOGICAL,
+					LogicalType:    LogicalOperationType(logicalOperation.LogOpType),
+					LeftOperation:  decodedOperations[logicalOperation.LeftOperationIndex],
+					RightOperation: decodedOperations[logicalOperation.RightOperationIndex],
+				})
+			} else if LogicalOperationType(logicalOperation.LogOpType) == OR {
+				decodedOperations = append(decodedOperations, &OrOperation{
+					OpType:         LOGICAL,
+					LogicalType:    LogicalOperationType(logicalOperation.LogOpType),
+					LeftOperation:  decodedOperations[logicalOperation.LeftOperationIndex],
+					RightOperation: decodedOperations[logicalOperation.RightOperationIndex],
+				})
+			} else {
+				return nil, errors.New("unknown logical operation type")
+			}
+		} else {
+			return nil, errors.New("unknown logical operation type")
+		}
+		log.Debug("Decoded operation", "operation", operation, "decodedOperations", decodedOperations)
+	}
+
+	var stack []Operation
+
+	for _, op := range decodedOperations {
```

Comment:
in typescript we have a while here right? https://github.com/river-build/river/blob/928170dd252f4d3a876bb2fab285b856ec3aa06a/packages/web3/src/entitlement.ts#L650

then we constantly overwrite the stack variable with the updated tree until it only has one root.

tbh I haven't been able to run that code in my head yet - I don't really get how the serialized data fits together with the operations tree, but I'm pretty sure this code is different and wanted to call that out. Good test cases should be able to uncover discrepancies.